### ISSUE TITLE
Overlap detection on Area2D is one frame late

### ISSUE THREAD

**Issue by Penguinum** [godotengine/godot#38983]
**Godot version:**
3.2.1-stable

**OS/device including version:**
Ubuntu 19.10

**Issue description:**
It looks like Area2D overlap is detected not in frame when it happens but in next frame. Maybe it's an expected behavior though it's not clear from docs.

**Steps to reproduce:**
Setup Area2D and KinematicBody2D moving into it. Call queue_free on KinematicBody2D when it enters Area2D. After that KinematicBody2D will still be visible in the next frame.
![Godot](https://user-images.githubusercontent.com/5759312/82731799-3f9a2a80-9d1a-11ea-96c2-abe91e44d374.gif)

**Minimal reproduction project:**
When you hold space bullets are spawned. They go from left to right and then enter Area2D. When they enter the Area2D queue_free is called for them but they are removed only after 2 frames in Area2D.
[RemoveBulletWhenItEntersArea2D.zip](https://github.com/godotengine/godot/files/4671813/RemoveBulletWhenItEntersArea2D.zip)

**Comment by Penguinum** [comment]
Update: in previous project get_overlapping_bodies was used, this could be the reason:
> For performance reasons (collisions are all processed at the same time) this list is modified once during the physics step, not immediately after objects are moved. Consider using signals instead.

However same happens in updated project when using body_entered signal.
[RemoveBulletWhenItEntersArea2D.zip](https://github.com/godotengine/godot/files/4671853/RemoveBulletWhenItEntersArea2D.zip)

**Comment by madmiraal** [comment]
This is not a bug, it's doing exactly what it's asked to do.

This is the sequence of events from the minimal reproduction project:
In each physics loop:
- Check for overlapping bodies and add them to the queue to be deleted.
- If space bar is pressed add new `Bullet`.
- Move existing `Bullets` to the right.

The three `Bullets` are therefore as describe in the above three steps:
Left: The new one that was created. (It will be moved in the next frames)
Middle: The one that has just been moved into the `Area` (it will be detected as overlapping in the next frame).
Right: The one that has been detected as overlapping, is in the queue to be deleted (it will be deleted before the next frame) and moved a second time.

**Comment by Penguinum** [comment]
@madmiraal, thank you for the reply.
> it will be detected as overlapping in the next frame
Does it mean `body_entered` is emitted next frame after `Bullet` is actually entered `Area`? Why?

**Comment by madmiraal** [comment]
> Does it mean body_entered is emitted next frame after Bullet is actually entered Area? Why?

Yes, collision detection (are the bodies overlapping) and state update (move the bodies) are done separately. This is to ensure that the order the bodies' states are updated in isn't important.

**Comment by Penguinum** [comment]
Thanks.

**Comment by MightyPrinny** [comment]
Signals are emitted after the physics step, so not in the next frame, buf if you use get_overlapping_bodies you'll get a list from the previous physics step. If you want to do something after the physics step the workaround is to have a rigid body in the root so it persist through the game that doesn't sleep, you can use its integrate forces callback to do a deferred call,because integrate forces runs during the physics simulation a deferred call will run at the end of the simulation, from there you can do a group call to notify your nodes at the end of the physics step.

**Comment by davidmcasas** [comment]
Frustratingly enough, this is still a thing in Godot 4.2.
Area2D body_entered signal is emitted a frame late. To be precise, a physics tick late, not a frame. Therefore, you can't rely on Area2D if you want to avoid the KinematicBody2D/CharacterBody2D overlapping the Area2D. Even if you made the bullet invisible on the body_entered callback with "body.visible = false", you'd still see it overlap.

I understand that the engine was designed this way, but what if I really NEED to make the body_entered be executed on the same tick? There seems to be no "clean" workaround to make the signals be emitted on the same frame that they are triggered. And this might be the reason I drop Godot altogether.

**Comment by danbrown1235813** [comment]
Agreed I just need a coin to disappear when something touches it, should be a core mechanic
